guides/source/working_with_javascript_in_rails.md

```diff
@@ -3,561 +3,197 @@
 
-Unobtrusive JavaScript
-----------------------
+Import maps
+————————————
 
-Rails uses a technique called "Unobtrusive JavaScript" to handle attaching
-JavaScript to the DOM. This is generally considered to be a best-practice
-within the frontend community, but you may occasionally read tutorials that
-demonstrate other ways.
+[Import maps](https://github.com/rails/importmap-rails) let you import JavaScript modules using logical names that map to versioned/digested files directly from the browser. Import maps are the new default for Rails 7, allowing anyone to build modern JavaScript applications using JavaScript libraries built for ESM without the need for transpiling or bundling.
 
-Here's the simplest way to write JavaScript. You may see it referred to as
-'inline JavaScript':
+Applications using import maps do not need Node or Yarn to function. If you plan to use Rails with importmap-rails to manage your JavaScript dependencies, you can use Rails without needing to install Node or Yarn.
 
-```html
-<a href="#" onclick="this.style.backgroundColor='#990000';event.preventDefault();">Paint it red</a>
-```
+When using import maps, no separate build step is required, just start your server with `rails s` and you are good to go.
 
-When clicked, the link background will become red. Here's the problem: what
-happens when we have lots of JavaScript we want to execute on a click?
-
-```html
-<a href="#" onclick="this.style.backgroundColor='#009900';this.style.color='#FFFFFF';event.preventDefault();">Paint it green</a>
-```
+### Adding NPM packages with importmap-rails
+To add new packages to your import map-powered application, run the `importmap pin` command from your terminal:
 
-Awkward, right? We could pull the function definition out of the click handler,
-and turn it into a function:
-
-```js
-window.paintIt = function(event, backgroundColor, textColor) {
-  event.preventDefault();
-  event.target.style.backgroundColor = backgroundColor;
-  if (textColor) {
-    event.target.style.color = textColor;
-  }
-}
+```bash
+./bin/importmap pin alpinejs
 ```
 
-And then on our page:
-
-```html
-<a href="#" onclick="paintIt(event, '#990000')">Paint it red</a>
+Then, import the package into `application.js` as usual:
+```javascript
+import Alpine from 'alpinejs'
 ```
 
-That's a little bit better, but what about multiple links that have the same
-effect?
+Bundlers
+————————————
+Import maps are the default for new Rails applications but if your application requires traditional JavaScript bundling, you can create new Rails applications with your choice of esbuild, Webpack, or rollup.js.
 
-```html
-<a href="#" onclick="paintIt(event, '#990000')">Paint it red</a>
-<a href="#" onclick="paintIt(event, '#009900', '#FFFFFF')">Paint it green</a>
-<a href="#" onclick="paintIt(event, '#000099', '#FFFFFF')">Paint it blue</a>
+To use a bundler instead of import maps in a new Rails application, pass the `—javascript` or `-j` option to `rails new`:
```

Comment:
Do you think it's worth explaining how to switch from import maps to a bundler if you need to? Even a really high level statement saying that it's possible might help.

Reply:
I think @rubys had some thoughts on how possible it really is...

Reply:
At the moment, I don't thing the guide should leave any impression that the choice you make now is one that you are stuck with for the life of the project.  It is not just a matter of switching from import maps to bundlers, but that turbo installs differently depending on that choice, as does stimulus, and (if you use either or both) does action cable and tailwind.  Your best best is to uninstall each in the proper order, and the reinstall them in the reverse order.  And there is no tooling support as of yet for uninstalling.  And that's just the gems that are maintained by the rails team - there may be third party gems with similar considerations.

I have no trouble with the `--api` and `--no-api` split.  If you know your client with be created with create-react-app, there are parts of Rails you won't ever need or use, so why install them?  If you decide later to add create-react-app, your may have parts of Rails you no longer use, but there is no urgent need to remove them.

The split between bundler and install map is not so clean.  I really would like a story to be something along the lines of unless you are absolutely certain you can't live without features like tree-shaking, make your life simpler by going with import maps.  I really do believe that you can build internet-scale apps using "only" Turbo, Stimulus, and Action Cable, and that Rails should be leading with that story.

But then there are things like Web Components which seems to be a rising star, and fits in well with a Rails import map application.  What concerns me is that much of that is moving towards TypeScript for the developer experience of things like IntelliSense in VIsual Studio Code, and the only support Rails currently has for TS is through bundlers.  So, unfortunately, there is a good case to be made for the position that if you might ever want something like that in the future, selecting a bundler at the start of the project is a good idea as it keeps this option open.  Even if you pick the wrong bundler - switching between bundlers is easier than switching from no bundler to bundler.

If there is interest in fixing any of this, I'll be glad to contribute code to the effort.

Reply:
Thanks for the thoughts on this @ghiculescu and @rubys

It sounds like we might be missing a section on which option to choose that:

- Leads with import maps as the default choice — this could be an appropriate place to paraphrase and link out to DHH's "3 great answers to JavaScript" post
- Shares the most prominent reasons you may wish to use a bundler instead (TypeScript, tree shaking, JS libraries that bundle CSS)
- Notes that any choice made when you create your application is not permanent without getting into the details of how to migrate from one to other. If you're moving from import maps to a bundler you can mostly get there from following the jsbundling install script errors but that doesn't seem well suited for this type of guide.

If this sounds right I can work on a first pass of that section later today.

Reply:
I agree with the first two bullets.  I feel that the third bullet would be leading people astray.  Before proceeding, I encourage you to try the following:

```
rails new throwaway
cd throwaway
bundle add jsbundling-rails 
./bin/rails javascript:install:esbuild
```

Observe the errors produced, and try to resolve.

Reply:
[Here's a gist](https://gist.github.com/DavidColby/3c176b70b9d8e3a450648361c5d05635) that covers everything needed to go from fresh Rails 7 + Hotwire + import maps to Hotwire + esbuild.

This obviously becomes more complex as you add dependencies. If it feels useful to demonstrate the basic steps for folks in the guide I'm happy to include the specific steps or an outline of what to do without the exact commands.

Reply:
I haven't reviewed this closely yet, but from a quick read: won't that leave things behind in assets/config/manifest.js?

We should compare notes.  My take on this: https://github.com/rubys/rails-uninstaller-tasks.  I believe that there needs to be more than a gist - there needs to be both a guide (separate from "getting started") and supporting tooling (my initial sketch was based on rails tasks, but there could be other means of getting this done).

Reply:
That's certainly possible.

I ran through quickly with the goal of booting up a default import maps powered application, confirming Turbo/Stimulus work in dev, ripping out import maps, adding esbuild, and getting back to the same dev state with Turbo/Stimulus working. There may be 1) things I'm missing in dev, particularly once a few external libs have been pinned and 2) production considerations that I've overlooked.

If you have time to review the gist a bit more and can provide feedback I'm all ears.

Stepping back a bit: for the purposes of this PR, I wonder if this detailed review of the switching steps needs to be included in the content.

I think it makes to include a note about the ability to switch along with a caution that switching may be time consuming if they need to do so at some point after `rails new`. We can encourage readers to think a bit about which option they choose while making it clear that it isn't impossible to migrate later, if they decide to.

The *how* of migrating feels better suited for a separate guide or, as you've proposed, a set of migration tasks that exist in importmap-rails/jsbundling-rails. At that point, adding a note to this guide that points the reader in the right direction could be added.

Reply:
> I think it makes to include a note about the ability to switch along with a caution that switching may be time consuming if they need to do so at some point after `rails new`. We can encourage readers to think a bit about which option they choose while making it clear that it isn't impossible to migrate later, if they decide to.

Yeah, I'm on board with this. Otherwise this could become a very long doc.

> The _how_ of migrating feels better suited for a separate guide or, as you've proposed, a set of migration tasks that exist in importmap-rails/jsbundling-rails. At that point, adding a note to this guide that points the reader in the right direction could be added.

I think this is probably the right guide for this to live in, but maybe this isn't the right PR for it.